Script Modules API

luisherranz opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/56313

---

EDIT: I've updated the description to showcase the latest changes.

This PR proposes a new Modules API for WordPress, designed to work with native ES Modules and Import Maps. It introduces functions such as `wp_register_script_module`, and `wp_enqueue_script_module`.

The API aims to provide a familiar experience to the existing `WP_Scripts` class, offering similar functionality. However, **it's not intended to duplicate the exact functionality of `WP_Scripts`;** rather, it is carefully tailored to address the specific needs and capabilities of ES modules.

For this initial version, **the current proposal is intentionally simplistic**, covering only the essential features needed to work with ES modules. Other enhancements and optimizations can be added later as the community identifies additional requirements and use cases.

### Differences Between WP_Script_Modules and WP_Scripts

1. **Dependency Specification**:

   With `WP_Script_Modules`, the array of dependencies supports not just strings but also arrays that include the dependency import type (`static` or `dynamic`). This design choice allows for future extensions of dependency properties, such as adding a `version` property to support "scopes" within import maps.

2. **Module Identifier**:

   Instead of a handle, `WP_Script_Modules` utilizes the module identifier, aligning with the module identifiers used in JavaScript files and import maps.

3. **Deregistration**:

   There is no equivalent of `wp_deregister_module` at this stage.

## API

### `wp_register_script_module( $module_identifier, $src, $deps, $version )`

Registers a module.

```php
// Registers a module with dependencies and versioning.
wp_register_script_module(
  'my-module',
  '/path/to/my-module.js',
  array( 'static-dependency-1', 'static-dependency-2' ),
  '1.2.3'
);
```

```js
// my-module.js
import { ... } from 'static-dependency-1';
import { ... } from 'static-dependency-2';

// ...
```

```php
// Registers a module with a dynamic dependency.
wp_register_script_module(
  'my-module',
  '/path/to/my-module.js',
  array(
    'static-dependency',
    array(
      'id'   => 'dynamic-dependency',
      'import' => 'dynamic'
    ),
  )
);
```

```js
// my-module.js
import { ... } from 'static-dependency';

// ...
const dynamicModule = await import('dynamic-dependency');
```

### `wp_enqueue_script_module( $module_identifier, $src, $deps, $version )`

Enqueues a module. If it includes a src, it will also register the module.

```php
wp_enqueue_script_module( 'my-module' );
```

### `wp_dequeue_script_module( $module_identifier )`

Dequeues a module.

```php
wp_dequeue_script_module( 'my-module' );
```

## Output

- When modules are enqueued, they are printed within script tags containing `type="module"` attributes. 
- Additionally, static dependencies of enqueued modules utilize `link` tags with `rel="modulepreload"` attributes.
- Lastly, an import map is generated and inserted using a `<script type="importmap">` tag.

```html
<script type="module" src="/path/to/my-module.js" id="my-module"></script>
<link rel="modulepreload" href="/path/to/static-dependency.js" id="static-dependency" />
<script type="importmap">
  {
    "imports": {
      "static-dependency": "/path/to/static-dependency.js",
      "dynamic-dependency": "/path/to/dynamic-dependency.js"
    }
  }
</script>
```

## Import Map Polyfill Requirement

Even though all the major browsers already support import maps, an import map polyfill is required until the percentage of users using old browser versions without import map support drops significantly. 

This work is ongoing and will be added to this PR once it's ready. Contributors can follow the progress in the associated GitHub issue: [guybedford/es-module-shims#406](https://github.com/guybedford/es-module-shims/issues/406).

## Feedback

This Modules API proposal is open for feedback and further discussion. Please contribute ideas, potential use cases, or any other thoughts that could make this feature robust and broadly useful for the WordPress community. Please use the [Trac Ticket](https://core.trac.wordpress.org/ticket/56313) or this pull request.

luisherranz committed:
Modules API: Initial version

luisherranz commented:
I see that PHP 8.3 has some deprecations when using `ReflectionProperty`. I'll review and fix those.

EDIT: I switched to `setStaticPropertyValue`/`getStaticPropertyValue` but now PHP 7.0 is failing. I'll investigate more 😄

EDIT 2: Fixed now in https://github.com/WordPress/wordpress-develop/pull/5818/commits/f13a068c13a694f3fc5fac34c9ec8374200a73a8!

swissspidy commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,299 @@
+ * @since 6.5.0
+ *
+ */
+add_action( 'wp_head', array( 'WP_Modules', 'print_import_map' ) );
```
These side effects should be moved to `default-filters.php` or into a function/method that can be called somewhere.

luisherranz replied:
Moved in https://github.com/WordPress/wordpress-develop/pull/5818/commits/27ba624094f23d0a15af4ce1d3fe8b11e679484b.

swissspidy commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,299 @@
+ * @param array             $dependencies      Optional. An array of module identifiers of the dependencies of this module. The dependencies can be strings or arrays. If they are arrays, they need an `id` key with the module identifier, and can contain a `type` key with either `static` or `dynamic`. By default, dependencies that don't contain a type are considered static.
+ * @param string|false|null $version           Optional. String specifying module version number. Defaults to false. It is added to the URL as a query string for cache busting purposes. If SCRIPT_DEBUG is true, the version is the current timestamp. If $version is set to false, the version number is the currently installed WordPress version. If $version is set to null, no version is added.
+ */
+function wp_register_module( $module_identifier, $src, $dependencies = array(), $version = false ) {
```
This file must only contain the `WP_Modules` class. Any other functions etc. should be in its own file.

luisherranz replied:
Moved to their own file in https://github.com/WordPress/wordpress-develop/pull/5818/commits/156eb588ec82771974f3722729b6c4b4685561b3.

swissspidy commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,299 @@
+	public static function print_import_map() {
+		$import_map = self::get_import_map();
+		if ( ! empty( $import_map['imports'] ) ) {
+			echo '<script type="importmap">' . wp_json_encode( self::get_import_map(), JSON_HEX_TAG | JSON_HEX_AMP ) . '</script>';
```
Can we use `wp_print_inline_script_tag` here?

luisherranz replied:
Changed in https://github.com/WordPress/wordpress-develop/pull/5818/commits/bbae17c72567a67a09487faffa73fe160218f904.

It adds a CDATA wrapper, but I guess that's ok, isn't it?

```html
<script type="importmap">
  /* <![CDATA[ */
  { "imports": { "dep": "\/dep.js?ver=6.5-alpha-56966-src" } }
  /* ]]> */
</script>
```

swissspidy commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,299 @@
+/**
+ * Core class used to register modules.
+ */
+class WP_Modules {
```
I know it's explained a little bit in the PR description, but it would be good to have a discussion about `WP_Dependencies` and `_WP_Dependency` and how these all fit together.

luisherranz replied:
Let's discuss that in the [ticket](https://core.trac.wordpress.org/ticket/56313).

swissspidy commented on tests/phpunit/tests/modules.php:
```diff
@@ -0,0 +1,429 @@
+<?php
+/**
+ * @group modules
+ * @covers ::wp_register_module
+ * @covers ::wp_enqueue_module
+ * @covers ::wp_dequeue_module
+ */
+class Tests_Modules extends WP_UnitTestCase {
```
Maybe split the tests here into two files, one for the functions and one for the class methods. There's definitely a lack of `@covers` annotations for all the methods being tested here, such as for `print_module_preloads` or `print_import_map`.

luisherranz replied:
Split and added `@covers` for the print methods in https://github.com/WordPress/wordpress-develop/pull/5818/commits/76e87268154ea5938c111bc12478bd37d8c45bc0.

luisherranz committed:
Move actions to default-filters.php

luisherranz committed:
Move module functions to its own file

luisherranz committed:
Use wp_print_inline_script_tag to print import map

luisherranz committed:
Divide tests into two files and improve covers

luisherranz committed:
Switch to ReflectionProperty to setStaticPropertyValue

luisherranz commented:
Thanks for the review, @swissspidy. I've made all the necessary changes 🙂

luisherranz committed:
Refactor the tests set_up

sirreal reviewed (commented):
I used this PR to build out what it would take to handle `viewModule` in `block.json` (analagous module version of `viewScript`).

One thing that I encountered is that module registration returns `null`. We don't have a return value that indicates success or failure. Look at this block:

https://github.com/WordPress/wordpress-develop/blob/455044e6261e64970fb9039dba52ceb47290a634/src/wp-includes/blocks.php#L190-L199

When implementing similar handling for modules, there's no return value from `wp_register_module` or `WP_Modules::register`, so we end up optimistically continuing as if the module is correctly registered.

Was that an intentional design decision? I'd expect to have some return value, maybe the module ID or an array that's descriptive of the module (`[ id => …, src => …, …]`) or `false` in case of failure.

luisherranz commented:
When I was studying WP Scripts, I saw that `wp_register_script` returns `false` when the script is already registered and `true` otherwise:

```php
if ( isset( $this->registered[ $handle ] ) ) {
  return false;
}
$this->registered[ $handle ] = new _WP_Dependency( $handle, $src, $deps, $ver, $args );
// ...
return true;
```

_[code](https://github.com/WordPress/wordpress-develop/blob/455044e6261e64970fb9039dba52ceb47290a634/src/wp-includes/class-wp-dependencies.php#L257-L259)_

So, it works like this:

- Returns `true`: the script has been successfully registered.
- Returns `false`: the script hasn't been registered again, but it was successfully registered before, so _it is_ successfully registered. 

In both cases, the script is successfully registered.

> so we end up optimistically continuing as if the module is correctly registered

That's my point. It doesn't matter what WP Scripts returns, the script is always registered so we should be able to continue as if the script is correctly registered.

I'm not opposed to mimicking the `true`/`false` return of WP Scripts, but I'd like to understand the use cases first in case there's an opportunity for improvement in the API 🙂

So, is there a use case where that `true`/`false` return is actually useful information?

sirreal commented:
Good point. I agree, it doesn't seem very helpful to distinguish between "registered on this call" and "already registered" if the result is the same. If there's no real failure scenario, it seems fine to always return `null`.

luisherranz commented:
Ok. Thanks, Jon. Let's keep it as it is for now then, just to keep open the possibility of returning something useful in the future or until we discover of use case for the WP Scripts `true`/`false` behavior 🙂

felixarntz commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,245 @@
+	 * @param array             $dependencies      Optional. An array of module identifiers of the dependencies of this module. The dependencies can be strings or arrays. If they are arrays, they need an `id` key with the module identifier, and can contain a `type` key with either `static` or `dynamic`. By default, dependencies that don't contain a type are considered static.
+	 * @param string|false|null $version           Optional. String specifying module version number. Defaults to false. It is added to the URL as a query string for cache busting purposes. If SCRIPT_DEBUG is true, the version is the current timestamp. If $version is set to false, the version number is the currently installed WordPress version. If $version is set to null, no version is added.
+	 */
+	public static function register( $module_identifier, $src, $dependencies = array(), $version = false ) {
```
it would be great to avoid all the static methods and properties and instead rely on instances. Static methods should only be used for things that are stateless, which isn't the case here, so in that situation using static methods is an antipattern, for example because it makes unit testing almost impossible.

I would recommend to make everything here non-static and instead introduce a simple `wp_modules()` function that encapsulates the default/main instance of the class, similar to how the existing `wp_scripts()` and `wp_styles()` functions do.

luisherranz replied:
Makes sense. Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/1aa45a6198e73869dcf4e642c3d2978888af612d.

felixarntz commented on src/wp-includes/modules.php:
```diff
@@ -0,0 +1,46 @@
+ * @param string|false|null $version           Optional. String specifying module version number. Defaults to false. It is added to the URL as a query string for cache busting purposes. If SCRIPT_DEBUG is true, the version is the current timestamp. If $version is set to false, the version number is the currently installed WordPress version. If $version is set to null, no version is added.
+ */
+function wp_register_module( $module_identifier, $src, $dependencies = array(), $version = false ) {
+	WP_Modules::register( $module_identifier, $src, $dependencies, $version );
```
See my above comment: If you go with a non-static class and a `wp_modules()` function, then this would still be simple and could become
```suggestion
	wp_modules()->register( $module_identifier, $src, $dependencies, $version );
```

Same applies to the other functions below.

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/1aa45a6198e73869dcf4e642c3d2978888af612d.

felixarntz commented on src/wp-includes/default-filters.php:
```diff
@@ -748,4 +748,9 @@
 add_action( 'wp_head', 'wp_print_font_faces', 50 );
 
+// Modules
+add_action( 'wp_head', array( 'WP_Modules', 'print_import_map' ) );
+add_action( 'wp_head', array( 'WP_Modules', 'print_enqueued_modules' ) );
+add_action( 'wp_head', array( 'WP_Modules', 'print_module_preloads' ) );
```
It would be great if you could introduce wrapper functions for these 3 as well (i.e. `wp_print_import_map()` etc.). This way you can reference the functions here, which avoids loading/instantiating the class early in the WordPress bootstrap process where it may not be needed.

Not a big deal, but an easy enough optimization IMO worth making.

luisherranz replied:
Sure. Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/eaaa563b214fd1bcd2b16642f6e628554711748b 🙂

felixarntz commented on tests/phpunit/tests/modules/modules.php:
```diff
@@ -0,0 +1,410 @@
+	public function set_up() {
+		parent::set_up();
+
+		$registered = new ReflectionProperty( 'WP_Modules', 'registered' );
+		$registered->setAccessible( true );
+		$this->old_registered = $registered->getValue();
+		$registered->setValue( null, array() );
+
+		$enqueued_before_registered = new ReflectionProperty( 'WP_Modules', 'enqueued_before_registered' );
+		$enqueued_before_registered->setAccessible( true );
+		$this->old_enqueued_before_registered = $enqueued_before_registered->getValue();
+		$enqueued_before_registered->setValue( null, array() );
```
See my above comment. This kind of hack is exactly why static classes for something stateful are an anti-pattern. By properly using a non-static class (i.e. a class instance), all of this could be replaced with a simple:
```suggestion
		$this->modules = new WP_Modules();
```

Then the tests below could all use `$this->modules`, each of which would be a clean instance.

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/1aa45a6198e73869dcf4e642c3d2978888af612d.

felixarntz commented on tests/phpunit/tests/modules/modules.php:
```diff
@@ -0,0 +1,410 @@
+		$enqueued_before_registered->setValue( null, array() );
+	}
+
+	public function tear_down() {
+		$registered = new ReflectionProperty( 'WP_Modules', 'registered' );
+		$registered->setAccessible( true );
+		$registered->setValue( null, $this->old_registered );
+
+		$enqueued_before_registered = new ReflectionProperty( 'WP_Modules', 'enqueued_before_registered' );
+		$enqueued_before_registered->setAccessible( true );
+		$enqueued_before_registered->setValue( null, $this->old_enqueued_before_registered );
+
+		parent::tear_down();
+	}
```
See above, by using instances instead of a static class, this entire hack/method could be removed.

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/1aa45a6198e73869dcf4e642c3d2978888af612d.

felixarntz reviewed (changes requested):
@luisherranz The logic here looks great to me.

I do however think the implementation should rely on a non-static class, as only stateless classes/methods should be static. The existing `WP_Scripts` and `WP_Styles` classes aren't static either, so that's another (albeit minor) reason.

felixarntz commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,245 @@
+ * Core class used to register modules.
+ */
+class WP_Modules {
+	/**
+	 * Holds the registered modules, keyed by module identifier.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @var array
+	 */
+	private static $registered = array();
+
+	/**
+	 * Holds the module identifiers that were enqueued before registered.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @var array
+	 */
+	private static $enqueued_before_registered = array();
```
Nit-pick: Property docs typically shouldn't have a blank line between those annotations.
```suggestion
	/**
	 * Holds the registered modules, keyed by module identifier.
	 *
	 * @since 6.5.0
	 * @var array
	 */
	private static $registered = array();

	/**
	 * Holds the module identifiers that were enqueued before registered.
	 *
	 * @since 6.5.0
	 * @var array
	 */
	private static $enqueued_before_registered = array();
```

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/e30c8e9ae0849f863bc730c131bd594edfc38ef8.

gziolo commented on tests/phpunit/tests/modules/modules.php:
```diff
@@ -0,0 +1,410 @@
+		$enqueued_modules = $this->get_enqueued_modules();
+
+		$this->assertEquals( 2, count( $enqueued_modules ) );
+		$this->assertEquals( true, str_starts_with( $enqueued_modules['foo'], '/foo.js' ) );
```
Nit, assertions like that can be simplified with different helper methods:
```suggestion
		$this->assertTrue( str_starts_with( $enqueued_modules['foo'], '/foo.js' ) );
```

or:

```suggestion
		$this-> assertStringStartsWith( '/foo.js', $enqueued_modules['foo'] );
```

More in https://docs.phpunit.de/en/9.6/assertions.html.

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/e644da8072e242f3cb118d43f71410252f90d4af.

gziolo commented on tests/phpunit/tests/modules/modules.php:
```diff
@@ -0,0 +1,410 @@
+
+		$enqueued_modules = $this->get_enqueued_modules();
+
+		$this->assertEquals( 1, count( $enqueued_modules ) );
```
Nit, possible alternative:
```suggestion
		$this-> assertCount( 1, $enqueued_modules );
```

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/e644da8072e242f3cb118d43f71410252f90d4af.

gziolo commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,245 @@
+ */
+
+/**
+ * Core class used to register modules.
```
Nit:
```suggestion
 * Core class used to register modules.
 *
 * @since 6.5.0
```

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/212cdac16228c427c7fdd6f142f5d6127ef9ac78.

gziolo commented:
> Good point. I agree, it doesn't seem very helpful to distinguish between "registered on this call" and "already registered" if the result is the same. If there's no real failure scenario, it seems fine to always return null.

In the case of `WP_Scripts`, it bails out early when there is the same handle registered, and therefore prevents overriding the same handle by plugins. That's why [`wp_deregister_script`](https://developer.wordpress.org/reference/functions/wp_deregister_script/) exists, so it was still possible. Example usage in the Gutenberg plugin:

https://github.com/WordPress/gutenberg/blob/d3071ca03bf3db4cd4267b4d9eb48c5658491c0f/lib/blocks.php#L219-L232

It's going to be challenging to replicate the same code with the currently proposed approach when the only possibility is to dequeue the module. In the shared example, all these scripts might not be enqueued as it happens later in the process. However, maybe it's fine, as when using the proposed implementation for modules, the plugin can register the same handles again, and they will get successfuly overridden. The only difference would be that some unused modules would remain registered and hopefully never enqueued.

`WP_Scripts` also offers [`wp_script_is`](https://developer.wordpress.org/reference/functions/wp_script_is/) that allows checking whether the script was already registered or enqueued. In that sense, it's similar to what the `false` return value does when calling `wp_register_script` that tries to register a previously registered script handle.

luisherranz commented:
@felixarntz: thanks! It makes sense. I'll make those changes 🙂

@gziolo: So the use case for `wp_deregister_script` is to prevent the script from being enqueued in the situation where someone else is still calling `wp_enqueue_script`?

gziolo commented:
>  So the use case for wp_deregister_script is to prevent the script from being enqueued in the situation where someone else is still calling wp_enqueue_script?

Yes, this is how it works. There is a check that ensures that only registered scripts get printed to make it possible to safely deregister a script.

luisherranz committed:
Test that static dependencies of dynamic dependencies are not preloaded

luisherranz committed:
Add since to main class DocBlock

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

luisherranz committed:
Change assertEquals with custom ones

felixarntz commented:
> `WP_Scripts` also offers [`wp_script_is`](https://developer.wordpress.org/reference/functions/wp_script_is/) that allows checking whether the script was already registered or enqueued. In that sense, it's similar to what the `false` return value does when calling `wp_register_script` that tries to register a previously registered script handle.

I agree providing an equivalent method (and wrapper function) to check whether a module has been registered or enqueued would be helpful. Either a `wp_module_is( $module, $status )` or two individual `wp_module_is_registered( $module )` and `wp_module_is_enqueued( $module )`.

Since this class implementation is technically separate from `WP_Scripts` and `WP_Styles`, I don't think we necessarily have to use the same approach for that. I don't think we explicitly support e.g. `queue` or `to_do` here anyway. FWIW I think the `wp_script_is()` function isn't great from an API perspective with requiring a "status" for something that should rather be distinct functions, so I'd prefer to go with two separate functions, one for registration check and one for enqueue check. Curious what you think.

In any case, if there are any concerns or pushback on this, it could also be added later; not a blocker from my perspective for the initial commit.

gziolo commented:
> In any case, if there are any concerns or pushback on this, it could also be added later; not a blocker from my perspective for the initial commit.

Agreed. Not a blocker for the initial version. Let’s see if `wp_module_is_registered()` and friends turn out to be necessary when working on refinements in the Gutenberg plugin to use the version we discuss here. The same applies to `wp_deregister_module`. We can always add the polyfill in the Gutenberg plugin if there is no other way to correctly override view modules that WordPress core will ship in the future.

swissspidy commented on tests/phpunit/tests/modules/wpModules.php:
```diff
@@ -0,0 +1,28 @@
+<?php
+/**
+ * @group modules
+ * @covers WP_Modules::get_version_query_string
```
```suggestion
 * @coversDefaultClass WP_Modules
```

luisherranz replied:
Done as part of https://github.com/WordPress/wordpress-develop/pull/5818/commits/1aa45a6198e73869dcf4e642c3d2978888af612d.

swissspidy commented on tests/phpunit/tests/modules/wpModules.php:
```diff
@@ -0,0 +1,28 @@
+	 *
+	 * @ticket 56313
+	 *
+	 * @covers WP_Modules::get_version_query_string
```
```suggestion
	 * @covers ::get_version_query_string
```

luisherranz replied:
Done as part of https://github.com/WordPress/wordpress-develop/pull/5818/commits/1aa45a6198e73869dcf4e642c3d2978888af612d.

swissspidy commented:
> * Additionally, static dependencies of enqueued modules utilize `link` tags with `rel="modulepreload"` attributes.

From the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/modulepreload):

> Preloading allows modules and their dependencies to be downloaded early, and can also significantly reduce the overall download and processing time. This is because it allows pages to fetch modules in parallel, instead of sequentially as each module is processed and its dependencies are discovered. _Note however that you can't just preload everything! What you choose to preload must be balanced against other operations that might then be starved, adversely affecting user experience._

Has any testing been done in this regard?

luisherranz committed:
Avoid static methods and rely on instances

luisherranz committed:
Introduce wrapper functions for printing methods

To avoid instantiating the class early in the WordPress bootstrap.

luisherranz committed:
Remove space between @since and @var in class properties

luisherranz commented:
> Has any testing been done in this regard?

@swissspidy, the static dependencies of the enqueued modules are always required to load those enqueued modules, so the current logic preloads those static dependencies, but not the dynamic ones.

```js
// foo.js
import bar from 'bar'; // static dependency, always used, required to load foo.js

const baz = await import( 'baz' ); // dynamic dependency, it may or may not be used
```

```html
<script type="importmap">
  { "imports": { "bar": "/bar.js", "baz": "/baz.js" } }
</script>

<!-- foo.js is enqueued -->
<script type="module" src="/foo.js"></script>

<!-- bar.js is preloaded, but baz.js is not -->
<link rel="modulepreload" src="/bar.js">
```

This also works recursively, so if `bar` has a static dependency, that dependency is also preloaded. This way, we ensure that **everything that is required on page load is downloaded in parallel**, without causing any waterfalls.

swissspidy commented:
Greg just brought up the topic of internationalization / localization. Has this been considered for this API here? I imagine people will want to use translations in modules. We'll probably need something similar to `wp_set_script_translations`, but also a solution that works well for dynamic dependencies.

gziolo commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,246 @@
+	 * @param string|false|null $version           Optional. String specifying module version number. Defaults to false. It is added to the URL as a query string for cache busting purposes. If SCRIPT_DEBUG is true, the version is the current timestamp. If $version is set to false, the version number is the currently installed WordPress version. If $version is set to null, no version is added.
+	 */
+	public function register( $module_identifier, $src, $dependencies = array(), $version = false ) {
+		if ( ! isset( $this->registered[ $module_identifier ] ) ) {
```
> Registers the module if no module with that module identifier has already been registered.

How can plugins override the registered module? It would be great to add a unit test that covers that use case, too.

luisherranz replied:
They currently can't. Do you think it's necessary for the initial version? 

If so, what would be the ideal API to override a module registration? 

I'm not familiar with the use case, but to me, the fact that you need to deregister the script to override it it's not very straightforward.

```php
if ( wp_script_is( 'foo', 'registered' ) ) {
  // Deregister so we can override.
  wp_deregister_script( 'foo' );
}
wp_register_script( 'foo', ... );
```

Or alternatively:

```php
$registered = wp_register_script( 'foo', ... );
if ( ! $registered ) {
  wp_deregister_script( 'foo' ); // Deregister so we can override.
  wp_register_script( 'foo', ... );
}
```

I'd rather expose a more explicit API, but curious to hear other's opinions 🙂

felixarntz replied:
+1 for only allowing to "override" by deregistering the module explicitly first. That is also the way required for the existing core APIs for scripts and stylesheets.

I think an explicit API makes it less error-prone, allowing to simply override could lead to accidental overrides too easily, while by requiring deregistration first it'll surely be intentional.

luisherranz replied:
I've been thinking about this during the last few days, and I don't think adding a deregister method is a good idea at this moment.

First, the use cases for that API are indirect:

- Override a module.
- Prevent the enqueuing of a module.

There is no clear direct use case for deregistering a module.

But the main reason is that there can be multiple sources for the same module identifier:

```html
<script type="importmap">
  {
    "imports": {
      "foo": "/foo.js"
    },
    "scopes": {
      "/bar.js": {
        "foo": "/my-own-foo.js"
      }
    }
  }
</script>
```

That's a very powerful feature that we should study how to leverage in the future.

If, at some point, we allow `wp_register_module` to register different modules for the same module identifier, it's not clear which one `wp_deregister_module` should deregister.

At least, I'd rather merge this initial version without it, and wait to see what other options we have to solve the use cases that are presented when we start integrating it.

<details>
 <summary>Some examples of how we could leverage `scopes` that come to my mind.</summary>

- Support different versions through semantic versioning

  ```php
  wp_register_script( 'foo', '/foo-v1.js', array(), '1.2.3' );
  wp_register_script( 'foo', '/foo-v2.js', array(), '2.4.5' );

  wp_register_script( 'bar', '/bar.js', array( 'foo' ) ); // Latest version.
  wp_register_script(
    'baz',
    '/baz.js',
    array( array( 'id' => 'foo', 'version' => '^1.0.0' ) )
  );
  ```

  ```html
  <script type="importmap">
    {
      "imports": {
        "foo": "/foo-v2.js"
      },
      "scopes": {
        "/baz.js": {
          "foo": "/foo-v1.js"
        }
      }
    }
  </script>
  ```

- Compat adapters for blocks using old versions

  ```json
  // old block
  {
    "apiVersion": 3
    // ...
  }
  ```

  ```json
  // new block
  {
    "apiVersion": 4
    // ...
  }
  ```

  ```html
  <script type="importmap">
    {
      "imports": {
        "@wordpress/block-editor": "/wp-includes/js/block-editor.js"
      },
      "scopes": {
        "/wp-content/plugins/old-block/edit.js": {
          "@wordpress/block-editor": "/wp-includes/js/block-editor-v3-compat.js"
        }
      }
    }
  </script>
  ```

- Translation APIs

  ```js
  // some-block/edit.js
  import { __ } from "@wordpress/i18n";
  ```

  ```html
  <script type="importmap">
    {
      "imports": {
        "@wordpress/i18n": "/wp-includes/js/i18n.js"
      },
      "scopes": {
        "/wp-content/plugins/some-block/edit.js": {
          "@wordpress/i18n": "/wp-includes/js/i18n.js?load=some-block"
        }
      }
    }
  </script>
  ```
</details>

gziolo commented on tests/phpunit/tests/modules/wpModules.php:
```diff
@@ -0,0 +1,445 @@
+	 * @covers ::get_version_query_string()
+	 */
+	public function test_get_version_query_string() {
+		$get_version_query_string = new ReflectionMethod( $this->modules, 'get_version_query_string' );
```
One more thing that isn't covered in the existing tests is asserting that the version argument passed during the module registration is correctly propagated to the printed import map, preloaded modules, and enqueued modules.

luisherranz replied:
True! Added in https://github.com/WordPress/wordpress-develop/pull/5818/commits/a13bc8758263e21587a23dbd9db378588ce0f24c 🙂

luisherranz committed:
Test that version is propagated correctly

felixarntz commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,246 @@
+	 *
+	 * @param string            $module_identifier The identifier of the module. Should be unique. It will be used in the final import map.
+	 * @param string            $src               Full URL of the module, or path of the script relative to the WordPress root directory.
+	 * @param array             $dependencies      Optional. An array of module identifiers of the dependencies of this module. The dependencies can be strings or arrays. If they are arrays, they need an `id` key with the module identifier, and can contain a `type` key with either `static` or `dynamic`. By default, dependencies that don't contain a type are considered static.
+	 * @param string|false|null $version           Optional. String specifying module version number. Defaults to false. It is added to the URL as a query string for cache busting purposes. If SCRIPT_DEBUG is true, the version is the current timestamp. If $version is set to false, the version number is the currently installed WordPress version. If $version is set to null, no version is added.
```
Can you break the parameter descriptions in multiple lines and cap them at a reasonable length? There is no set standard for that I believe, but typically PHP projects use either 80 characters or 120 characters per line.

Mainly I just want to avoid these excessive line lengths. When broken into separate lines, make sure to indent them (with spaces) to start at the same column as the initial line description.

luisherranz replied:
Sure. Capped to 120 characters in https://github.com/WordPress/wordpress-develop/pull/5818/commits/67f3900835d224d931b7183ea4fef5ddedc59387.

felixarntz commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,246 @@
+	public function enqueue( $module_identifier ) {
+		if ( isset( $this->registered[ $module_identifier ] ) ) {
+			$this->registered[ $module_identifier ]['enqueued'] = true;
+		} elseif ( ! in_array( $module_identifier, $this->enqueued_before_registered, true ) ) {
+			$this->enqueued_before_registered[] = $module_identifier;
```
I would recommend to make the `$enqueued_before_registered` property a map of `$module_identifier => true`. The value wouldn't carry any meaning, but it's simply a small performance optimization, as it would allow us to use `isset()` instead of `in_array()` / `array_search()`, which is faster. Dequeuing would be faster too, as it could use `unset()` instead of `array_splice()`.

For instance here this would become:
```suggestion
		} elseif ( ! isset( $this->enqueued_before_registered[ $module_identifier ] ) ) {
			$this->enqueued_before_registered[ $module_identifier ] = true;
```

You could even avoid the `elseif` since unconditionally setting it to `true` would have the same outcome (as it's now an associative array where the same module couldn't appear multiple times).
```suggestion
		} else {
			$this->enqueued_before_registered[ $module_identifier ] = true;
```

luisherranz replied:
Makes sense, thank you 🙂

Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/66987f0f5df80620e1f6982313e6916b322f941c.

felixarntz commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,246 @@
+		if ( isset( $this->registered[ $module_identifier ] ) ) {
+			$this->registered[ $module_identifier ]['enqueued'] = false;
+		}
+		$key = array_search( $module_identifier, $this->enqueued_before_registered, true );
+		if ( false !== $key ) {
+			array_splice( $this->enqueued_before_registered, $key, 1 );
+		}
```
See my above comment, this could be simplified to:
```suggestion
		unset( $this->enqueued_before_registered[ $module_identifier ] );
```
No check needed, as `unset()` with a missing key just does nothing, no warnings/notices either.

luisherranz replied:
Done as part of https://github.com/WordPress/wordpress-develop/pull/5818/commits/66987f0f5df80620e1f6982313e6916b322f941c.

felixarntz commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,246 @@
+			$this->registered[ $module_identifier ] = array(
+				'src'          => $src,
+				'version'      => $version,
+				'enqueued'     => in_array( $module_identifier, $this->enqueued_before_registered, true ),
```
See my other comment, this could be slightly faster as:
```suggestion
				'enqueued'     => isset( $this->enqueued_before_registered[ $module_identifier ] ),
```

luisherranz replied:
Done as part of https://github.com/WordPress/wordpress-develop/pull/5818/commits/66987f0f5df80620e1f6982313e6916b322f941c.

felixarntz commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,246 @@
+	 * @return string A string with the version, prepended by `?ver=`, or an empty string if there is no version.
+	 */
+	private function get_version_query_string( $version ) {
+		if ( defined( 'SCRIPT_DEBUG ' ) && SCRIPT_DEBUG ) {
+			return '?ver=' . time();
```
Hmm, this isn't a thing in the regular script API. While I'm not necessarily opposed to it, I'm a bit wary of introducing it without a bit more discussion. For example:
* If this is a good idea, then it should also be done in the regular `WP_Scripts` class.
* Maybe there are other concerns though, e.g. related to sites that use `SCRIPT_DEBUG` in production. That's certainly not a good idea, but for sites that do, this behavior would make their performance a lot worse by avoiding caching.

Again, I think it may be a good idea, but I think we should validate that first with other stakeholders. Maybe leave it out for the initial iteration?

luisherranz replied:
I agree, let's leave it out for the initial version. We can add it later or keep it on higher-level abstractions, [like what Gutenberg does](https://github.com/WordPress/gutenberg/blob/e8b0a4c459dd2dd8f6775f312128cc6e7106a6bc/lib/blocks.php#L250).

Removed in https://github.com/WordPress/wordpress-develop/pull/5818/commits/fda294d858731120c1651dbec70aa3468030f726.

felixarntz reviewed (commented):
Looks almost good to go from my perspective! Only very minor feedback at this point.

luisherranz committed:
Cap parameter descriptions to 120 chars

luisherranz committed:
Turn $enqueued_before_registered into a map

luisherranz committed:
Remove SCRIPT_DEBUG support

felixarntz reviewed (approved):
Thank you for the iterations @luisherranz, looks great to me!

swissspidy commented:
@felixarntz I think we'll want to have a story for localization first , see https://github.com/WordPress/wordpress-develop/pull/5818#issuecomment-1876870705

felixarntz commented:
That should definitely be a priority, but IMO not a blocker for a first commit.

luisherranz commented on src/wp-includes/default-filters.php:
```diff
@@ -748,4 +748,9 @@
 add_action( 'wp_head', 'wp_print_font_faces', 50 );
 
+// Modules.
+add_action( 'wp_head', 'wp_print_import_map' );
+add_action( 'wp_head', 'wp_print_enqueued_modules' );
+add_action( 'wp_head', 'wp_print_module_preloads' );
```
Note: we need to move these prints to the footer on classic themes ([related Gutenberg PR](https://github.com/WordPress/gutenberg/pull/57396)).

luisherranz replied:
[The solution used in Gutenberg](https://github.com/WordPress/gutenberg/pull/57396) was just temporary because there's no need to print all the modules (or preloads) at once. So taking into account that the order in which modules are printed is not important, and that the sooner they are on the page the better because the browser will start downloading them sooner:

- I've modified the print functions which now keep track of the printed modules and preloads and therefore they can be called multiple times.
- I've added hooks to print both the enqueued and preloaded modules at both the `wp_head` and `wp_footer`. They will print the modules that are already available in the `wp_head` (like modules of blocks in block themes, or modules from classic themes or plugins) and they will print again the modules that were not available in the `wp_head` but are available in the `wp_footer` (like modules of blocks in classic themes).
- I've moved the import map to the footer to make sure that it includes all the dependencies.

This is the explanation included in the code:

```php
/**
 * Adds the hooks to print the import map, enqueued modules and module
 * preloads.
 *
 * It adds the actions to print the enqueued modules and module preloads to
 * both `wp_head` and `wp_footer` because in classic themes, the modules
 * used by the theme and plugins will likely be able to be printed in the
 * `head`, but the ones used by the blocks will need to be enqueued in the
 * `footer`.
 *
 * As all modules are deferred and dependencies are handled by the browser,
 * the order of the modules is not important, but it's still better to print
 * the ones that are available when the `wp_head` is rendered, so the browser
 * starts downloading those as soon as possible.
 *
 * The import map is also printed in the footer to be able to include the
 * dependencies of all the modules, including the ones printed in the footer.
 *
```

luisherranz replied:
@c4rl0sbr4v0 discovered that we can't enqueue or preload any module before we know all the dependencies because the import map needs to be added before any module:

- https://github.com/WordPress/gutenberg/pull/57778#discussion_r1450185051

To be honest, this is quite a surprise, especially the preload part. If someone knows better, please tell so.

I've already opened a Trac ticket to fix it: https://core.trac.wordpress.org/ticket/60240

luisherranz committed:
Rename enqueued to enqueue

westonruter commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,251 @@
+ *
+ * @since 6.5.0
+ */
+class WP_Modules {
```
IMO, `WP_Script_Modules` would be a clearer name. The `WP_Modules` name is very generic and "module" could refer to anything, like `WP_Widget`.

luisherranz replied:
Let's discuss this [in the PR itself](https://github.com/WordPress/wordpress-develop/pull/5818#issuecomment-1883520368) 🙂

westonruter commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,251 @@
+		if ( ! isset( $this->registered[ $module_identifier ] ) ) {
+			$deps = array();
+			foreach ( $dependencies as $dependency ) {
+				if ( isset( $dependency['id'] ) ) {
```
Either this or in an `else` (see below) at the end would perhaps be more robust to help authors.

```suggestion
				if ( is_array( $dependency ) ) {
					if ( ! isset( $dependency['id'] ) ) {
						_doing_it_wrong( __METHOD__, __( 'Missing required id key in entry among dependencies array.' ), '6.5.0' );
						continue;
					}
```

luisherranz replied:
Makes sense, thank you. Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/b3554d61c10ab2010d66178eca87d7dee62a470e.

westonruter commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,251 @@
+						'id'   => $dependency,
+						'type' => 'static',
+					);
+				}
```
See above.

```suggestion
				} else {
					_doing_it_wrong( __METHOD__, __( 'Entries in dependencies array must be either strings or arrays with an id key.' ), '6.5.0' );
				}
```

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/b3554d61c10ab2010d66178eca87d7dee62a470e.

westonruter commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,251 @@
+	 *                                             final import map.
+	 * @param string            $src               Full URL of the module, or path of the script relative to the WordPress
+	 *                                             root directory.
+	 * @param array             $dependencies      Optional. An array of module identifiers of the dependencies of this
```
Perhaps this could be more explicit? Granted, using array shapes in core is not common, although they are used in `src/wp-includes/sodium_compat`:

```suggestion
	 * @param array<string|array{id: string, type?: 'static'|'dynamic' }>             $dependencies      Optional. An array of module identifiers of the dependencies of this
```

luisherranz replied:
That's better, thanks. Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/df73485c9c43995eea8b298138df43e258cef0d4.

westonruter commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,251 @@
+			wp_print_script_tag(
+				array(
+					'type' => 'module',
+					'src'  => $module['src'] . $this->get_version_query_string( $module['version'] ),
```
While not common, it's possible that the `src` here may already have a query string (such as when importing a script from a service on another origin). Therfore, I think `add_query_arg` should rather be used here. The method can be changed to return an array:

```suggestion
					'src'  => add_query_arg( $this->get_version_query_arg( $module['version'] ), $module['src'] ),
```

westonruter replied:
Or even better, there should be a new method introduced like `get_versioned_src( $module )` which returns the string instead of relying on concatenation in 3 places here. So this would be:

```suggestion
					'src'  => $this->get_versioned_src( $module ),
```

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/1df672d3da05ae7bfdf52caffac3a9d4d38752e1.

westonruter commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,251 @@
+	public function get_import_map() {
+		$imports = array();
+		foreach ( $this->get_dependencies( array_keys( $this->get_enqueued() ) ) as $module_identifier => $module ) {
+			$imports[ $module_identifier ] = $module['src'] . $this->get_version_query_string( $module['version'] );
```
```suggestion
			$imports[ $module_identifier ] = $module['src'] . $this->get_versioned_src( $module );
```

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/1df672d3da05ae7bfdf52caffac3a9d4d38752e1.

westonruter commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,251 @@
+			if ( true !== $module['enqueued'] ) {
+				echo sprintf(
+					'<link rel="modulepreload" href="%s" id="%s">',
+					esc_attr( $module['src'] . $this->get_version_query_string( $module['version'] ) ),
```
See above, and nit `s/esc_attr/esc_url/`

```suggestion
					esc_url( $this->get_versioned_src( $module ) ),
```

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/0cb7f1b53dac51b1b0933744604676c0e265b30d and https://github.com/WordPress/wordpress-develop/pull/5818/commits/1df672d3da05ae7bfdf52caffac3a9d4d38752e1.

westonruter commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,251 @@
+				echo sprintf(
+					'<link rel="modulepreload" href="%s" id="%s">',
+					esc_attr( $module['src'] . $this->get_version_query_string( $module['version'] ) ),
+					esc_attr( $module_identifier )
```
I think there should be a prefix or suffix added here since there could be conflicts with other elements on the page. The `WP_Scripts` class adds a suffix.

```suggestion
					esc_attr( $module_identifier . '-js-modulepreload' )
```

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/834693d4a2223a52bb3f78ab410c8a867108aa5d.

westonruter commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,251 @@
+		}
+	}
+
+	/**
+	 * Gets the version of a module.
+	 *
+	 * If $version is set to false, the version number is the currently installed
+	 * WordPress version. If $version is set to null, no version is added.
+	 * Otherwise, the string passed in $version is used.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string|false|null $version The version of the module.
+	 * @return string A string with the version, prepended by `?ver=`, or an empty string if there is no version.
+	 */
+	private function get_version_query_string( $version ) {
+		if ( false === $version ) {
+			return '?ver=' . get_bloginfo( 'version' );
+		} elseif ( null !== $version ) {
+			return '?ver=' . $version;
+		}
+		return '';
+	}
```
```suggestion
	/**
	 * Gets the versioned URL for a module src.
	 *
	 * If $version is set to false, the version number is the currently installed
	 * WordPress version. If $version is set to null, no version is added.
	 * Otherwise, the string passed in $version is used.
	 *
	 * @since 6.5.0
	 *
	 * @param array $module The module.
	 * @return string The module src with a version if relevant.
	 */
	private function get_versioned_src( array $module ) {
		$args = array();
		if ( false === $module['version'] ) {
			$args['ver'] = get_bloginfo( 'version' );
		} elseif ( null !== $module['version'] ) {
			$args['ver'] = $module['version'];
		}
		if ( $args ) {
			return add_query_arg( $args, $module['src'] );
		}
		return $module['src'];
	}
```

luisherranz replied:
Makes sense. Thanks, Weston!

Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/1df672d3da05ae7bfdf52caffac3a9d4d38752e1.

westonruter commented on src/wp-includes/modules.php:
```diff
@@ -0,0 +1,104 @@
+ *                                             final import map.
+ * @param string            $src               Full URL of the module, or path of the script relative to the WordPress
+ *                                             root directory.
+ * @param array             $dependencies      Optional. An array of module identifiers of the dependencies of this
```
See note above about possibly incorporating a more strict PHP type with an array shape.

luisherranz replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/df73485c9c43995eea8b298138df43e258cef0d4.

westonruter commented on src/wp-includes/class-wp-modules.php:
```diff
@@ -0,0 +1,251 @@
+			wp_print_inline_script_tag(
+				wp_json_encode( $import_map, JSON_HEX_TAG | JSON_HEX_AMP ),
+				array(
+					'type' => 'importmap',
```
Adding an `id` here would seem useful as WP generally adds IDs to other scripts bring printed to the page.

```suggestion
					'type' => 'importmap',
 					'id'   => 'wp-js-module-importmap',

```

luisherranz replied:
Done as part of https://github.com/WordPress/wordpress-develop/pull/5818/commits/834693d4a2223a52bb3f78ab410c8a867108aa5d.

westonruter reviewed (changes requested):
Overall looks great! Looking forward to having this in core. 

Just a few suggestions for improvement or discussion.

luisherranz committed:
Add failing test for multiple prints of enqueued modules

luisherranz committed:
Fix test

luisherranz committed:
Add failing test for multiple prints of preloaded modules

luisherranz committed:
Fix the test

luisherranz committed:
Print enqueued and preloaded modules in the head and footer

luisherranz committed:
Move the hooks to the class

luisherranz committed:
Rename `type` key to `import`

It's semantically more correct and it liberates
the `type` key to be used by other purposes,
like specifying between `script`, `css` or
`json`.

luisherranz committed:
Add _doing_it_wrongs

Co-authored-by: Weston Ruter <westonruter@google.com>

luisherranz committed:
Merge remote-tracking branch 'origin/add/modules-api' into add/modules-api

luisherranz committed:
Improve @param type of $deps

luisherranz committed:
Indicate how the new print functions work

luisherranz committed:
Reorder methods

luisherranz committed:
Test an empty import map

luisherranz committed:
Replace get_version_query_string with get_versioned_src

luisherranz committed:
Use `esc_url` instead of `esc_attr` for link's href attribute

luisherranz committed:
Add suffixes to the ids and an id to the import map

luisherranz commented:
Thanks for the thorough review, Weston 🙂

I've addressed all the feedback except the one about renaming the class to `WP_Script_Modules`, which I like to discuss a bit more before doing so.

It's very likely that, in the future, once the import attributes are final, we'll add support for importing CSS and JSON:

```html
<script type="importmap">
  { "imports": { "some-styles-package": "/wp-content/.../styles.css" } }
</script>
```

```js
import sheet from "some-styles-package" with { type: "css" };
document.adoptedStyleSheets = [sheet];
```

Preloads are a bit different and they can't be enqueued on their own (so that means we would need to allow people to specify the `type`), but other than that, they are pretty much like any other module.

It seems like JSON modules are officially referred as **JSON modules** ([link](https://github.com/tc39/proposal-json-modules)), whereas CSS modules are officially referred as **CSS module scripts** ([link]()) because there was [a discussion to rename it from **CSS modules**](https://github.com/WICG/webcomponents/issues/843) to avoid confusion with the existing [CSS Modules project](https://github.com/css-modules/css-modules). 

So, I'm not saying that `WP_Script_Modules` is worse than `WP_Modules`, but that we should also think about the possibility of powering types of modules that are not scripts, and a class name that is suitable for all of them:

- (Java)Script modules
- CSS module scripts
- JSON modules

What do you think?

westonruter commented:
@luisherranz Interesting. I wasn't aware of JSON and CSS also being on the roadmap for being modules. Nevertheless, I think `WP_Script_Modules` is still appropriate for them as well, as you listed they all mention "script":

* Java**Script** modules
* CSS module **scripts**
* JSON (Java**Script** Object Notation) modules

Additionally, all of these modules alike end up getting printed in an importmap `script`, and more importantly, they all get used in a _script_ context. So they are all inherently "scripty".

luisherranz committed:
Rename class to WP_Script_Modules

luisherranz committed:
Allow wp_enqueue_module to also register modules

luisherranz commented:
Ok! Thanks, Weston. I've renamed the class to `WP_Script_Modules`.

I've also added support to register the script with a single call to the enqueue function, as requested by @aaronjorbin in [the ticket](https://core.trac.wordpress.org/ticket/56313#comment:33). 

I still agree that it would have made sense to separate both registration and enqueuing as a way to teach people that those are two separate steps (especially when working with modules where registered modules can't only end up being enqueued, but also preloaded or added to the import map), but as @aaronjorbin pointed out, the `wp_enqueue_style` is still going to be relevant in the future, and it would be confusing to have different APIs for both:

```php
wp_enqueue_style( 'my-style', '/styles.css' );
wp_register_module( 'my-module', '/module.js' );
wp_enqueue_module( 'my-module' );
```

So now, this also works:

```php
wp_enqueue_style( 'my-style', '/styles.css' );
wp_enqueue_module( 'my-module', '/module.js' );
```

---

From my point of view, I think this is in a great state. I don't have plans to add any other changes to this initial version.

westonruter reviewed (approved):
LGTM!

ockham commented:
Now that the `WP_Modules` class has been renamed to `WP_Script_Modules` -- should the `wp_modules()` function also be renamed to `wp_script_modules()` (for consistency)?

@luisherranz @westonruter

luisherranz commented:
I'd keep it as `wp_modules` to keep consistency with the register and enqueue functions.

ockham commented:
> I'd keep it as `wp_modules` to keep consistency with the register and enqueue functions.

Works for me! And if we end up deciding otherwise, we can always change it in a follow-up 😊

ockham commented:
Committed to Core in https://core.trac.wordpress.org/changeset/57269.

Thank you, folks! This is big 🎉

luisherranz commented:
Follow-up tickets:

- **Import map polyfill**: https://core.trac.wordpress.org/ticket/60232
- **Support for `editorModule`/`module`/`viewModule` in `block.json`**: https://core.trac.wordpress.org/ticket/60233
- **Translations API**: https://core.trac.wordpress.org/ticket/60234

mukeshpanchal27 commented on src/wp-includes/class-wp-script-modules.php:
```diff
@@ -0,0 +1,352 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param array $module_ids The identifiers of the modules for which to gather dependencies.
+	 * @param array $import_types       Optional. Import types of dependencies to retrieve: 'static', 'dynamic', or both.
```
Indention issue.

```
* @param array $module_ids   The identifiers of the modules for which to gather dependencies.
* @param array $import_types Optional. Import types of dependencies to retrieve: 'static', 'dynamic', or both.
```

Nit-pick

luisherranz replied:
Fixed by @ockham [here](https://core.trac.wordpress.org/changeset/57271) (thanks Bernie!).

felixarntz commented:
It's great to see this committed! 🎉 

Nevertheless, I think the naming topic was wrapped a bit quickly as it was only initiated 2 days ago by @westonruter, so that didn't leave much room for other people's thoughts before the commit. We can of course follow up if needed, but a final decision on naming needs to happen soon before more other things are implemented on top of the API.

I think we should continue the discussion, either here or in a new issue. I personally find it a confusing choice to go with `WP_Script_Modules` class but `wp_*_module()` function names. What's the rationale to do it differently between those two places? To me it doesn't seem to make sense.
* Either we go with the more verbose script modules, in expectation ("future proof") that other kinds of modules will be made available via WordPress APIs in the future.
* Or we stick to just "module" and agree on that, when we talk about "modules" in WordPress context, it's always JS modules.

I don't have a strong opinion which way to go, but I think the current solution is neither.

westonruter commented:
IMO, I think the functions should be `wp_script_modules()`, `wp_register_script_module()`, and `wp_enqueue_script_module()` for consistency.

gziolo commented:
> IMO, I think the functions should be `wp_script_modules()`, `wp_register_script_module()`, and `wp_enqueue_script_module()` for consistency.

It makes perfect sense after the class name changes were applied.

azaozz commented:
> I think the functions should be wp_script_modules(), wp_register_script_module(), and wp_enqueue_script_module() for consistency.

Another +1 for this.

luisherranz commented:
Seems unanimous, then. I'll prepare a PR with the renaming.

luisherranz commented:
There you go:

- https://github.com/WordPress/wordpress-develop/pull/5869
